programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+                        whence = SEEK_END;
+                        break;
+                    default:
+                        ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                        break;
+                }
+
+                if ( ret == 0 )
+                    ret = fseek( file, offset, whence );
+            }
+        }
+        break;
+    case MBEDTLS_SERIALIZE_FUNCTION_FTELL:
+        {
+            FILE * file = NULL;
+            int32_t file_id;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            file_id = item_int32( inputs[0] );
+            file = (FILE *)get_file_from_id( file_id );
+
+            if ( file )
+            {
+                ret = ftell( file );
+
+                if ( ret >= 0 )
+                {
+                    ALLOC_OUTPUT( 0, sizeof( int32_t ) );
+                    set_item_int32( outputs[0], ret );
+                    ret = 0;
+                }
+            }
+        }
+        break;
+    case MBEDTLS_SERIALIZE_FUNCTION_FERROR:
+        {
+            FILE * file = NULL;
+            int32_t file_id;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            file_id = item_int32( inputs[0] );
+            file = (FILE *)get_file_from_id( file_id );
+
+            if ( file )
+            {
+                ret = ferror( file );
+            }
+        }
+        break;
+    case MBEDTLS_SERIALIZE_FUNCTION_DOPEN:
+        {
+            char * path = NULL;
+            int32_t file_id;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            path = item_buffer( inputs[0] );
+            DBG( "open dir [%s]", path);
+            file_id = alloc_file_context();
+            DBG( "allocated dir id [%d]", file_id);
+            if ( file_id != -1 )
+            {
+                DIR * dir = opendir( path );
+                if ( dir != NULL )
+                {
+                    files[file_id - 1].file = (void *)dir;
+                    ALLOC_OUTPUT( 0, sizeof (int32_t) );
+                    set_item_int32( outputs[0], file_id );
+                    ret = 0;
+                }
+                else
+                {
+                    DBG( "opendir: error = %s", strerror( errno ) );
+                    free_file_context( file_id );
+                }
+            }
+        }
+        break;
+    case MBEDTLS_SERIALIZE_FUNCTION_DREAD:
+        {
+            int32_t file_id, size;
+            DIR * dir = NULL;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            size = item_int32( inputs[0] );
+            file_id = item_int32( inputs[1] );
+            dir = (DIR *)get_file_from_id( file_id );
+            if ( dir )
+            {
+                struct dirent * entry = readdir( dir );
+                if ( entry )
+                {
+                    int file_size = strlen( entry->d_name );
+
+                    /* Check buffer size */
+                    if ( file_size < size )
```

Comment:
There are many error paths here and they all get lumped into the default `MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT` error code. This will make debugging harder (we don't use `errno` in our file I/O code in the library, but some example programs do, and it's useful to query it in a debugger). In addition, as mentioned above, `BAD_OUTPUT` is not a correct error code here.